src/Lykke.Service.History/Workflow/ExecutionProcessing/BaseBatchQueueReader.cs

```diff
@@ -0,0 +1,180 @@
+﻿using System;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Threading;
+using System.Threading.Tasks;
+using Common.Log;
+using Lykke.Common.Log;
+using RabbitMQ.Client;
+using RabbitMQ.Client.Events;
+
+namespace Lykke.Service.History.Workflow.ExecutionProcessing
+{
+    public abstract class BaseBatchQueueReader<T> : IDisposable
+    {
+        private const int ReconnectTimeoutSeconds = 30;
+
+        private readonly string _connectionString;
+        private readonly int _prefetchCount;
+        private readonly int _batchCount;
+        private CancellationTokenSource _cancellationTokenSource;
+        private Task _queueReaderTask;
+
+        protected readonly ILog Log;
+        protected ConcurrentQueue<CustomQueueItem<T>> Queue;
+
+        protected BaseBatchQueueReader(
+            ILogFactory logFactory,
+            string connectionString,
+            int prefetchCount,
+            int batchCount)
+        {
+            _connectionString = connectionString;
+            _prefetchCount = prefetchCount;
+            _batchCount = batchCount;
+            Log = logFactory.CreateLog(this);
+        }
+
+        public void Dispose()
+        {
+            Stop();
+        }
+
+        public void Start()
+        {
+            _cancellationTokenSource = new CancellationTokenSource();
+            Queue = new ConcurrentQueue<CustomQueueItem<T>>();
+            _queueReaderTask = StartRabbitSessionWithReconnects();
+        }
+
+        public void Stop()
+        {
+            _cancellationTokenSource.Cancel();
+
+            Task.WaitAny(_queueReaderTask, Task.Delay(10000));
+        }
+
+        protected abstract string ExchangeName { get; }
+
+        protected abstract string QueueName { get; }
+
+        protected abstract string[] RoutingKeys { get; }
+
+        protected abstract EventHandler<BasicDeliverEventArgs> CreateOnMessageReceivedEventHandler(IModel channel);
+
+        protected abstract Task ProcessBatch(IList<CustomQueueItem<T>> batch);
+
+        private async Task StartRabbitSessionWithReconnects()
+        {
+            while (!_cancellationTokenSource.IsCancellationRequested)
+            {
+                try
+                {
+                    await StartQueueReader();
+                }
+                catch (Exception ex)
```

Comment:
Maybe we can use specified exception?

Reply:
I don't know what type of exception will be risen here.

Reply:
We should handle any exception, because we don't want to exit `while` if it is not requested.